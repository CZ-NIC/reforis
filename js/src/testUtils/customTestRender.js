/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
