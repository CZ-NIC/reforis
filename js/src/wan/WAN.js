/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import update from 'immutability-helper';
import propTypes from 'prop-types';

import ForisForm from '../formContainer/ForisForm';
import {APIEndpoints} from '../common/API';

import WAN6Form, {validateWAN6Form} from './WAN6Form';
import MACForm, {validateMACForm} from './MACForm';
import WANForm, {validateWANForm} from './WANForm';
import ConnectionTest from '../connectionTest/ConnectionTest';

WAN.propTypes = {
    ws: propTypes.object.isRequired
};

export default function WAN({ws}) {
    return <>
        <ForisForm
            ws={ws}
            forisConfig={{
                endpoint: APIEndpoints.wan,
                wsModule: 'wan'
            }}
            prepData={prepData}
            prepDataToSubmit={prepDataToSubmit}
            validator={validator}
        >
            <WANForm/>
            <WAN6Form/>
            <MACForm/>
        </ForisForm>

        <h1>{_('Connection test')}</h1>
        <p
            dangerouslySetInnerHTML={{
                __html: _(`
Here you can test you connection settings. Remember to click on the <b>Save button</b> before running the test.
Note that sometimes it takes a while before the connection is fully initialized. So it might be useful to
wait for a while before running this test.
        `)
            }}
        />
        <ConnectionTest ws={ws} type='wan'/>
    </>
}

function prepData(formData) {
    // Create empty form fields if nothing has got from the server.
    const wan6_6in4 = update((formData.wan6_settings || {}).wan6_6in4 || {}, {
        dynamic_ipv4: {
            $set: ((formData.wan6_settings || {}).wan6_6in4 || {}).dynamic_ipv4 || {enabled: false}
        },
    });
    return update(formData, {
        wan_settings: {
            wan_dhcp: {$set: formData.wan_settings.wan_dhcp || {}},
            wan_static: {$set: formData.wan_settings.wan_static || {}},
            wan_pppoe: {$set: formData.wan_settings.wan_pppoe || {}}
        },
        wan6_settings: {
            wan6_dhcpv6: {$set: formData.wan6_settings.wan6_dhcpv6 || {duid: ""}},
            wan6_static: {$set: formData.wan6_settings.wan6_static || {}},
            wan6_6to4: {$set: formData.wan6_settings.wan6_6to4 || {}},
            wan6_6in4: {$set: wan6_6in4},
        }
    })
}


function prepDataToSubmit(formData) {
    const dataToSubmit = {
        wan_settings: deleteUnnecessarySettings(formData.wan_settings.wan_type, formData.wan_settings),
        wan6_settings: deleteUnnecessarySettings(formData.wan6_settings.wan6_type, formData.wan6_settings),
        mac_settings: formData.mac_settings,
    };

    if (formData.wan6_settings.wan6_type === '6in4' && !formData.wan6_settings.wan6_6in4.dynamic_ipv4.enabled)
        formData.wan6_settings.wan6_6in4.dynamic_ipv4 = {enabled: false};

    if (!formData.mac_settings.custom_mac_enabled)
        delete dataToSubmit.mac_settings.custom_mac;

    return dataToSubmit;
}

function deleteUnnecessarySettings(type, settings) {
    for (let key in settings) {
        if (!settings.hasOwnProperty(key) || key.endsWith('type'))
            continue;
        if (!key.endsWith(type))
            delete settings[key]
    }
    return settings
}

function validator(formData) {
    const errors = {
        wan_settings: validateWANForm(formData.wan_settings),
        wan6_settings: validateWAN6Form(formData.wan6_settings),
        mac_settings: validateMACForm(formData.mac_settings),
    };
    if (errors.wan_settings || errors.wan6_settings || errors.mac_settings)
        return errors;
    return null;
}
