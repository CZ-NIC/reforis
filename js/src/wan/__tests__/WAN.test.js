/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {fireEvent, render, waitForElement} from 'customTestRender';
import diffSnapshot from 'snapshot-diff';
import {mockedWS} from 'mockWS';
import {wanSettingsFixture} from './__fixtures__/wanSettings';
import mockAxios from 'jest-mock-axios';

import WAN from '../WAN';


describe('<WAN/>', () => {
    let asFragment;
    let firstRender;
    let getByLabelText;

    beforeEach(async () => {
        const mockWebSockets = new mockedWS();
        const renderRes = render(<WAN ws={mockWebSockets}/>);
        mockAxios.mockResponse({data: wanSettingsFixture()});
        asFragment = renderRes.asFragment;
        getByLabelText = renderRes.getByLabelText;

        await waitForElement(() => renderRes.getByText('WAN IPv4'));
        firstRender = renderRes.asFragment();
    });

    it('Snapshot: WAN IPv4 dhcp, WAN IPv6 dhcp, MAC disabled', () => {
        expect(firstRender).toMatchSnapshot();
    });

    it('Snapshot WAN IPv4 (static).', () => {
        fireEvent.change(getByLabelText('IPv4 protocol'), {target: {value: 'static'}});
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it('Snapshot WAN IPv4 (pppoe).', () => {
        fireEvent.change(getByLabelText('IPv4 protocol'), {target: {value: 'pppoe'}});
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it('Snapshot WAN IPv6 (dhcpv6).', () => {
        fireEvent.change(getByLabelText('IPv6 protocol'), {target: {value: 'dhcpv6'}});
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it('Snapshot WAN IPv6 (static).', () => {
        fireEvent.change(getByLabelText('IPv6 protocol'), {target: {value: 'static'}});
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it('Snapshot WAN IPv6 (6to4).', () => {
        fireEvent.change(getByLabelText('IPv6 protocol'), {target: {value: '6to4'}});
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it('Snapshot WAN IPv6 (6in4).', () => {
        fireEvent.change(getByLabelText('IPv6 protocol'), {target: {value: '6in4'}});
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it('Snapshot WAN IPv6 (6in4), dynamic IPv4 handling enabled.', () => {
        fireEvent.change(getByLabelText('IPv6 protocol'), {target: {value: '6in4'}});
        firstRender = asFragment();

        fireEvent.click(getByLabelText('Dynamic IPv4 handling'));
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it('Snapshot: MAC enabled', () => {
        fireEvent.click(getByLabelText('Custom MAC address'));
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });
});
