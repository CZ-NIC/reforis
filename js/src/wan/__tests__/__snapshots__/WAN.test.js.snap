// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WAN/> Snapshot WAN IPv4 (pppoe). 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -47,30 +47,69 @@
        </div>
        <div
          class=\\"form-group\\"
        >
          <label
-           for=\\"2\\"
+           for=\\"6\\"
          >
-           DHCP hostname
+           PAP/CHAP username
          </label>
          <div
            class=\\"input-group\\"
          >
            <input
-             class=\\"form-control\\"
-             id=\\"2\\"
+             class=\\"form-control  is-invalid\\"
+             id=\\"6\\"
+             required=\\"\\"
              type=\\"text\\"
              value=\\"\\"
            />
          </div>
-         <small
-           class=\\"form-text text-muted\\"
+         <div
+           class=\\"invalid-feedback\\"
          >
-           Hostname which will be provided to DHCP server.
-         </small>
+           This field is required.
+         </div>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"7\\"
+         >
+           PAP/CHAP password
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             autocomplete=\\"new-password\\"
+             class=\\"form-control  is-invalid\\"
+             id=\\"7\\"
+             required=\\"\\"
+             type=\\"password\\"
+             value=\\"\\"
+           />
+           <div
+             class=\\"input-group-append\\"
+           >
+             <button
+               class=\\"input-group-text\\"
+               type=\\"button\\"
+             >
+               <i
+                 class=\\"fa fa-eye\\"
+               />
+             </button>
+           </div>
          </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+       </div>
        <h3>
          WAN IPv6
        </h3>
        <div
          class=\\"form-group\\"
@@ -141,10 +180,11 @@
        <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-lg-2\\"
+           disabled=\\"\\"
            type=\\"submit\\"
          >
              Save
          </button>
        </div>"
`;

exports[`<WAN/> Snapshot WAN IPv4 (static). 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -47,28 +47,127 @@
        </div>
        <div
          class=\\"form-group\\"
        >
          <label
-           for=\\"2\\"
+           for=\\"6\\"
+         >
+           IP address
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"6\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+       </div>
+       <div
+         class=\\"form-group\\"
        >
-           DHCP hostname
+         <label
+           for=\\"7\\"
+         >
+           Network mask
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"7\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"8\\"
+         >
+           Gateway
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"8\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"9\\"
+         >
+           DNS server 1
          </label>
          <div
            class=\\"input-group\\"
          >
            <input
              class=\\"form-control\\"
-             id=\\"2\\"
+             id=\\"9\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           DNS server address is not required as the built-in DNS resolver is capable of working without it.
+         </small>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"10\\"
+         >
+           DNS server 2
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control\\"
+             id=\\"10\\"
              type=\\"text\\"
              value=\\"\\"
            />
          </div>
          <small
            class=\\"form-text text-muted\\"
          >
-           Hostname which will be provided to DHCP server.
+           DNS server address is not required as the built-in DNS resolver is capable of working without it.
          </small>
        </div>
        <h3>
          WAN IPv6
        </h3>
@@ -141,10 +240,11 @@
        <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-lg-2\\"
+           disabled=\\"\\"
            type=\\"submit\\"
          >
              Save
          </button>
        </div>"
`;

exports[`<WAN/> Snapshot WAN IPv6 (6in4), dynamic IPv4 handling enabled. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -242,10 +242,100 @@
                Some tunnel providers allow you to have public dynamic IPv4. Note that you need to fill in some extra fields to make it work.
              </small>
            </label>
          </div>
        </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"20\\"
+         >
+           Tunnel ID
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"20\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           ID of your tunnel which was assigned to you by the provider.
+         </small>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"21\\"
+         >
+           Username
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"21\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           Username which will be used to provide credentials to your tunnel provider.
+         </small>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"22\\"
+         >
+           Key
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"22\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           Key which will be used to provide credentials to your tunnel provider.
+         </small>
+       </div>
        <h3>
          MAC
        </h3>
        <div
          class=\\"form-group\\""
`;

exports[`<WAN/> Snapshot WAN IPv6 (6in4). 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -109,10 +109,143 @@
            >
              6in4 (public IPv4 address required)
            </option>
          </select>
        </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"8\\"
+         >
+           Provider IPv4
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"8\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           This address will be used as a endpoint of the tunnel on the provider's side.
+         </small>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"9\\"
+         >
+           Routed IPv6 prefix
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"9\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           IPv6 addresses which will be routed to your network.
+         </small>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"10\\"
+         >
+           MTU
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"10\\"
+             max=\\"1500\\"
+             min=\\"1280\\"
+             required=\\"\\"
+             type=\\"number\\"
+             value=\\"\\"
+           />
+           <div
+             class=\\"input-group-append\\"
+           >
+             <button
+               aria-label=\\"Increase\\"
+               class=\\"btn btn-outline-secondary\\"
+               type=\\"button\\"
+             >
+               <i
+                 class=\\"fas fa-plus\\"
+               />
+             </button>
+             <button
+               aria-label=\\"Decrease\\"
+               class=\\"btn btn-outline-secondary\\"
+               type=\\"button\\"
+             >
+               <i
+                 class=\\"fas fa-minus\\"
+               />
+             </button>
+           </div>
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <div
+           class=\\"custom-control custom-checkbox \\"
+         >
+           <input
+             class=\\"custom-control-input\\"
+             id=\\"11\\"
+             type=\\"checkbox\\"
+           />
+           <label
+             class=\\"custom-control-label\\"
+             for=\\"11\\"
+           >
+             Dynamic IPv4 handling
+             <small
+               class=\\"form-text text-muted\\"
+             >
+               Some tunnel providers allow you to have public dynamic IPv4. Note that you need to fill in some extra fields to make it work.
+             </small>
+           </label>
+         </div>
+       </div>
        <h3>
          MAC
        </h3>
        <div
          class=\\"form-group\\"
@@ -141,10 +274,11 @@
        <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-lg-2\\"
+           disabled=\\"\\"
            type=\\"submit\\"
          >
              Save
          </button>
        </div>"
`;

exports[`<WAN/> Snapshot WAN IPv6 (6to4). 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -109,10 +109,40 @@
            >
              6in4 (public IPv4 address required)
            </option>
          </select>
        </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"8\\"
+         >
+           Public IPv4
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"8\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           Public IPv4 address is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           In order to use 6to4 protocol, you might need to specify your public IPv4 address manually (e.g. when your WAN interface has a private address which is mapped to public IP).
+         </small>
+       </div>
        <h3>
          MAC
        </h3>
        <div
          class=\\"form-group\\"
@@ -141,10 +171,11 @@
        <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-lg-2\\"
+           disabled=\\"\\"
            type=\\"submit\\"
          >
              Save
          </button>
        </div>"
`;

exports[`<WAN/> Snapshot WAN IPv6 (dhcpv6). 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -109,10 +109,34 @@
            >
              6in4 (public IPv4 address required)
            </option>
          </select>
        </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"8\\"
+         >
+           Custom DUID
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control\\"
+             id=\\"8\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           DUID which will be provided to the DHCPv6 server.
+         </small>
+       </div>
        <h3>
          MAC
        </h3>
        <div
          class=\\"form-group\\""
`;

exports[`<WAN/> Snapshot WAN IPv6 (static). 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -109,10 +109,143 @@
            >
              6in4 (public IPv4 address required)
            </option>
          </select>
        </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"8\\"
+         >
+           Address
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"8\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           IPv6 address and prefix length for WAN interface e.g. 2001:db8:be13:37da::1/64
+         </small>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"9\\"
+         >
+           Gateway
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"9\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"10\\"
+         >
+           Prefix
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"10\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           Address range for local network, e.g. 2001:db8:be13:37da::/64
+         </small>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"11\\"
+         >
+           DNS server 1
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control\\"
+             id=\\"11\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           DNS server address is not required as the built-in DNS resolver is capable of working without it.
+         </small>
+       </div>
+       <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"12\\"
+         >
+           DNS server 2
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control\\"
+             id=\\"12\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           DNS server address is not required as the built-in DNS resolver is capable of working without it.
+         </small>
+       </div>
        <h3>
          MAC
        </h3>
        <div
          class=\\"form-group\\"
@@ -141,10 +274,11 @@
        <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-lg-2\\"
+           disabled=\\"\\"
            type=\\"submit\\"
          >
              Save
          </button>
        </div>"
`;

exports[`<WAN/> Snapshot: MAC enabled 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -137,14 +137,45 @@
              </small>
            </label>
          </div>
        </div>
        <div
+         class=\\"form-group\\"
+       >
+         <label
+           for=\\"9\\"
+         >
+           MAC address
+         </label>
+         <div
+           class=\\"input-group\\"
+         >
+           <input
+             class=\\"form-control  is-invalid\\"
+             id=\\"9\\"
+             required=\\"\\"
+             type=\\"text\\"
+             value=\\"\\"
+           />
+         </div>
+         <div
+           class=\\"invalid-feedback\\"
+         >
+           This field is required.
+         </div>
+         <small
+           class=\\"form-text text-muted\\"
+         >
+           Colon is used as a separator, for example 00:11:22:33:44:55
+         </small>
+       </div>
+       <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-lg-2\\"
+           disabled=\\"\\"
            type=\\"submit\\"
          >
              Save
          </button>
        </div>"
`;

exports[`<WAN/> Snapshot: WAN IPv4 dhcp, WAN IPv6 dhcp, MAC disabled 1`] = `
<DocumentFragment>
  <h1>
    WAN
  </h1>
  <p>
    
Here you specify your WAN port settings. Usually, you can leave these options untouched unless instructed
otherwise by your internet service provider. Also, in case there is a cable or DSL modem connecting your
router to the network, it is usually not necessary to change this setting.
        
  </p>
  <div
    class="col-sm-12 col-lg-12 p-0 mb-3"
  >
    <form>
      <h3>
        WAN IPv4
      </h3>
      <div
        class="form-group"
      >
        <label
          for="1"
        >
          IPv4 protocol
        </label>
        <select
          class="custom-select"
          id="1"
        >
          <option
            value="dhcp"
          >
            DHCP (automatic configuration)
          </option>
          <option
            value="static"
          >
            Static IP address (manual configuration)
          </option>
          <option
            value="pppoe"
          >
            PPPoE (for DSL bridges, Modem Turris, etc.)
          </option>
        </select>
      </div>
      <div
        class="form-group"
      >
        <label
          for="2"
        >
          DHCP hostname
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            id="2"
            type="text"
            value=""
          />
        </div>
        <small
          class="form-text text-muted"
        >
          Hostname which will be provided to DHCP server.
        </small>
      </div>
      <h3>
        WAN IPv6
      </h3>
      <div
        class="form-group"
      >
        <label
          for="3"
        >
          IPv6 protocol
        </label>
        <select
          class="custom-select"
          id="3"
        >
          <option
            value="none"
          >
            Disable IPv6
          </option>
          <option
            value="dhcpv6"
          >
            DHCPv6 (automatic configuration)
          </option>
          <option
            value="static"
          >
            Static IP address (manual configuration)
          </option>
          <option
            value="6to4"
          >
            6to4 (public IPv4 address required)
          </option>
          <option
            value="6in4"
          >
            6in4 (public IPv4 address required)
          </option>
        </select>
      </div>
      <h3>
        MAC
      </h3>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox "
        >
          <input
            class="custom-control-input"
            id="4"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="4"
          >
            Custom MAC address
            <small
              class="form-text text-muted"
            >
              Useful in cases, when a specific MAC address is required by your internet service provider.
            </small>
          </label>
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-lg-2"
          type="submit"
        >
            Save
        </button>
      </div>
    </form>
  </div>
  <h1>
    Connection Test
  </h1>
  <p>
    
Here you can test you connection settings. Remember to click on the 
    <b>
      Save button
    </b>
     before running the test.
Note that sometimes it takes a while before the connection is fully initialized. So it might be useful to
wait for a while before running this test.
        
  </p>
  <form>
    <table
      class="table table-borderless table-hover offset-lg-3 col-lg-6 col-sm-12"
    >
      <tbody>
        <tr>
          <th
            scope="row"
          >
            IPv4 connectivity
          </th>
          <td
            class="text-center"
          >
            <div
              class="text-secondary"
            >
              <i
                class="fas fa-minus"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv4 gateway connectivity
          </th>
          <td
            class="text-center"
          >
            <div
              class="text-secondary"
            >
              <i
                class="fas fa-minus"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv6 connectivity
          </th>
          <td
            class="text-center"
          >
            <div
              class="text-secondary"
            >
              <i
                class="fas fa-minus"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv6 gateway connectivity
          </th>
          <td
            class="text-center"
          >
            <div
              class="text-secondary"
            >
              <i
                class="fas fa-minus"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="col-sm-12 col-lg-12 p-0 mb-3 text-right"
    >
      <button
        class="btn btn-primary col-sm-12 col-lg-2"
        type="submit"
      >
          Test connection
      </button>
    </div>
  </form>
</DocumentFragment>
`;
